### PR TITLE
Design doc updates

### LINKED ISSUE #467: design doc: component configuration
Needs to cover in some detail:
- [x] vic-machine configuration of VCH
- [x] VCH component access to VCH config
- [x] configuration of containerVMs (overlap with #469)
- [x] tether access to containerVM config

Should touch on:
- [x] hidden configurations

### PR DESCRIPTION
**Pull request by hickeng**
This adds design doc for the component configuration mechanism (#229) and
reorganised what was there into a more easily consumable structure

Fixes #467

### CODE REVIEW

File: doc/design/components.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
+# vSphere Integrated Containers - Architecture and Design
+
+## Component Architecture
+
+This is a component architecture for the system, encompassing some of the main control flows. It is intended to be used as a visual map for which components interact and how those interaction occur; it does not indicate what those interaction are.
```
Comment (jzt) [inline doc/design/components.md]: `s/interaction/interactions`

File: doc/design/components.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+The component follows the naming pattern introduced by OCI with 'runc' and is a docker registry client library, purely concerned with the pull/push aspects of the registry with login becoming a transitive dependency.
+
+[Current open issues relating to imagec](https://github.com/vmware/vic/labels/component%2Fimagec)
+
+
+### Port Layer
+#### Port Layer - Execution
+
+This component deals handles management of containers such as create, start, stop, kill, etc, and is broken distinct from interaction primarily because the uptime requirements may be different, [see Interaction](#portlayer---interaction).
```
Comment (jzt) [inline doc/design/components.md]: `s/deals//`

Comment (jzt) [inline doc/design/components.md]: `and is broken distinct from interaction` seems to need fixing up.

File: doc/design/components.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+[Issues relating to portlayer execution](https://github.com/vmware/vic/labels/component%2Fportlayer%2Fexecution)
+
+
+#### Port Layer - Interaction
+
+This component deals handles interaction with a running container and is broken distinct from execution primarily because the uptime requirements may be different.
```
Comment (jzt) [inline doc/design/components.md]: This is a repeat of line 90 :)

File: doc/design/components.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+* [a custom init binary](#tether) that provides the command & control channel for container interaction
+
+The contianer process runs as root with full privileges, however there's no way to make persistent changes to anything except the container filesystem - the core operating files are read-only on the ISO and refresh each time the container is started.
+
+[Issues relating to this component](https://github.com/vmware/vic/labels/component%2Fcontainer-base)
+
+
+### Tether
+
+The tether is an init replacement used in containerVMs that provides the command & control channel necessary to perform any operation inside the container. This includes launching of the container process, setting of environment variables, configuration of networking, etc. The tether is currently based on a modified SSH server tailed specifically for this purpose.
```
Comment (dougm) [inline doc/design/components.md]: s/tailed/tailored/ ?

**Comment by dougm** [comment]
excellent :+1:

File: doc/design/components.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
+# vSphere Integrated Containers - Architecture and Design
+
+## Approach
+
+The devleopment approach taken for VIC is iterative refinement, both for the high level product and recursively throughout the lower level components that go into it. This, as with most Agile processes,is driven by a need to have a useful product in short timeframes, and to manage risk - whether risk related to larger delivery deadlines (VIC may be OSS but it's still a VMware product), changing technical landscapes, or the unknown. The day-by-day development process we follow is [documented in the contributor docs](../../CONTRIBUTING.md).
```
Comment (cgtexmex) [inline doc/design/components.md]: development misspelled

File: doc/design/communications.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
+# Communications
+
+This document discusses the communication channels that are possible between components, but not what flows over them. This is essentially Layer2/3 and Layer7 is discussed in the documents relating to specific components that talk over these channels.
```
Comment (mlh78750) [inline doc/design/communications.md]: Change : the communication channels that are possible between components
to: the communication channels supported between components

File: doc/design/communications.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+1. application level network traffic
+2. VIC control channel, which includes user interaction with the container console
+
+This section addresses the VIC control channel
+
+
+### Async - guestinfo
+The containerVM guestinfo is used as a persistent, asynch publishing mechanism for specific data related to container state, primarily:
+
+1. assigned DHCP address
```
Comment (mlh78750) [inline doc/design/communications.md]: Could this also be used for static addresses? Should we just say IP address?

File: doc/design/components.md [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+[Issues relating to vicadmin](https://github.com/vmware/vic/labels/component%2Fvicadmin)
+
+
+### Docker API server
+
+This is the portion of a Virtual Container Host that provides a Docker API endpoint for users to interact with; it is also referred to as a 'personality'. The longer term design has multiple personalities running within a single VCH, such that the same endpoint can serve mixed API versions, and the same VCH can serve multiple API flavours.
+
+As of the v0.1 release this makes use of the [docker engine-api](https://github.com/docker/engine-api) project to ensure API compatibility with docker.
+
+The following is commentary about why the docker API server is not actual Docker. It's not what would usually be found in design doc, but is a question that comes up enough that it becomes core to understanding the design.
```
Comment (cgtexmex) [inline doc/design/components.md]: need to add an "a" -- "It's not what would usually be found in **a** design doc, but is a question that comes up enough that it becomes core to understanding the design."

File: doc/design/communications.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+2. error reporting for [containerVM components](#components.md#container)
+3. exit status of the container process
+
+The other direction, appliance to containerVM, is addressed in a little more depth under [configuration](configuration.md#appliance)
```
Comment (mlh78750) [inline doc/design/communications.md]: Confused about "the other direction" reference here. Other than what? Are you saying that the info you have described so far is just the stuff that the container is putting in there and that the VCH "sets" will be described in configuration.md?

File: doc/design/components.md [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+[Current open issues relating to imagec](https://github.com/vmware/vic/labels/component%2Fimagec)
+
+
+### Port Layer
+
+The port layer presents container primitive operations, with APIs that express intent rather than the intricies of how.
+
+#### Port Layer - Execution
+
+This component deals handles management of containers such as create, start, stop, kill, etc, and is broken distinct from interaction primarily because the uptime requirements may be different, [see Interaction](#portlayer---interaction).
```
Comment (cgtexmex) [inline doc/design/components.md]: Remove deals from first sentence -- "This component handles..." should suffice

File: doc/design/components.md [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+If the execution portion of the port layer is unavailable then only container management operations around creating, starting, stopping, et al are impacted, and then only for as long as the VCH is unavailable.
+
+If the interaction portions are unavailable it impacts ongoing use of interactive sessions and potentially loses chunks of the container output (unless serialized to vSphere infrastructure as an intermediate step - [container logging](#container-logging), [container base](#container-base), and [tether](#tether)) are the other components that factor into the log persistence discussion).
+
+[Issues relating to portlayer interaction](https://github.com/vmware/vic/labels/component%2Fportlayer%2Finteraction)
+
+
+#### Port Layer - Networking
+
+A Virtual Container Host presents a cluster of resource as a single endpoint, so while we must also supply transparent cross host networking, it must be there without requiring any user configuration.
```
Comment (cgtexmex) [inline doc/design/components.md]: plural for resource "A Virtual Container Host presents a cluster of resource**s** as a single endpoint..."

File: doc/design/communications.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+Pros:
+* appliance does not require connectivity with the management network (when combined with [vmomi agent](#components.md#vmomi-authenticating-agent) )
+
+Cons:
+* require knowledge of applianceVM host and reconfiguration if applianceVM vMotion occurs
+
+#### vSPC relay
+A vSPC is explicitly to enable vMotion for VMs that have network serial connections. When connecting to a vSPC the connection is an extended TELNET protocol rather than raw TCP. In the case of a regular vSPC appliance, the vSPC has to bridge between the management network (network serial originates from the ESX host, not the VM) and a network the VIC appliance is attached to.
+The vSPC relay is an agent on ESX that does dumb forwarding of the incoming Telnet connection via vSocket into the VCH appliance, which provides a very basic vSPC implementation. In this case the containerVM would be configured with a network serial port with a vSPC target that is the ESX host running the applianceVM.
```
Comment (mlh78750) [inline doc/design/communications.md]: Maybe bump this bullet up above the first one. It sets some context that helps understand the first bullet.

File: doc/design/components.md (hunk 5 above)
Comment (mlh78750) [inline doc/design/components.md]: s/processes,is/processes, is/

File: doc/design/components.md [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+[Issues relating to portlayer storage](https://github.com/vmware/vic/labels/component%2Fvmomi%2Fgateway)
+
+
+
+## Install and management
+
+### vic-machine
+
+The _docker-machine_ mechanism is something of a de-facto standard for provisioning docker hosts.
+There is a significant amount of vSphere specific behaviour that needs to be expose, and that may well go beyond what docker-machine provides - hence supplying a vic-machine binary.
```
Comment (cgtexmex) [inline doc/design/components.md]: s/expose/exposed/

File: doc/design/components.md [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+
+[Issues relating to validating proxy](https://github.com/vmware/vic/labels/component%2Fvalidating-proxy)
+
+
+
+## ESX Agents
+
+### VMOMI Authenticating Agent
+
+As with [access to NSX management](#nsx-authenticating-agent), vSphere orchestration requires credentials and therefore credential management. The authenticating agent is indended to move the credential management problem out of the VCH and into the vSphere infrastructure. Authorizing the VCH to perform specific operations by virtue of being a specific VM (and potentially cryptographically verified), rather than generating and embedding credentials into the guest, means:
```
Comment (cgtexmex) [inline doc/design/components.md]: s/indended/intended/

File: doc/design/components.md [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+
+## Approach
+
+The devleopment approach taken for VIC is iterative refinement, both for the high level product and recursively throughout the lower level components that go into it. This, as with most Agile processes,is driven by a need to have a useful product in short timeframes, and to manage risk - whether risk related to larger delivery deadlines (VIC may be OSS but it's still a VMware product), changing technical landscapes, or the unknown. The day-by-day development process we follow is [documented in the contributor docs](../../CONTRIBUTING.md).
+
+The upshot of this approach is that many of these components are outline sketches at best as we start out, with specific areas being filled out over time. Avoiding any comment on business priorities, the _technical_ priorities for the project are:
+
+1. build an appliance foundation to permit secure orchestration of Virtual Infrastructure by regular users
+2. build primitves that support container style semantics, implemented with VMs
+2. present the Docker API as a language with which users can perform the orchestration they require
```
Comment (mlh78750) [inline doc/design/components.md]: Consider... present a Docker personality provides a Docker API endpoint that enables users to orchestrate the system described in the first two bullets using the Docker API.

File: doc/design/components.md [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+
+The tether is an init replacement used in containerVMs that provides the command & control channel necessary to perform any operation inside the container. This includes launching of the container process, setting of environment variables, configuration of networking, etc. The tether is currently based on a modified SSH server tailed specifically for this purpose.
+
+[Additional detail](tether.md)  
+[Issues relating to tether](https://github.com/vmware/vic/labels/component%2Ftether)
+
+
+### Container Logging
+
+Container logging, as distinct from other logging mechanisms, is the capture mechanism for container output. This fills the same niche as [Docker log drivers](https://docs.docker.com/engine/admin/logging/overview/) and has the potential to be a direct consumption of docker log drivers.
```
Comment (mlh78750) [inline doc/design/components.md]: a direct consumer --instead of direct consumption?

File: doc/design/vic-machine.md [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+  - IPAM config for _client_ network (defaults to DHCP)
+* datastore paths for (may all be the same path):
+  - images
+  - containers
+  - volumes
+
+
+## Implementation approach
+
+My current approach is to have several components that get rolled into both vic-machine. Initial implementation should focus on:
```
Comment (cgtexmex) [inline doc/design/vic-machine.md]: s/My/The/ ?

File: doc/design/vic-machine.md [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+* CLI parsing into config structs
+* validation of config struct values
+* reification of VCH from config struct
+
+Components:
+* command line argument parsing
+  - maps command line arguments to internal config data structures
+* validation of config structure values
+  - given an internal config data structure it's necessary to check the specifics against the target vSphere to ensure they are valid - past validity is no indicator of current validity
+  - this may also translate from symbolic names to morefs, resulting in a manifest that is resilient to name changes, but fragile across vSphere instances with identical naming schemes. This should be a user selectable behaviour, I prefer it on by default.
```
Comment (cgtexmex) [inline doc/design/vic-machine.md]: s/behaviour/behavior/ 

"I prefer it on by default" -- is this George speaking or the project? I'd recommend rewording that phrase or removing it and leaving "This should be user selectable behavior."

File: doc/design/vic-machine.md [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+* reification of configuration
+  - when a manifest has been validated against a vSphere, this component creates the corresponding vSphere objects - this is primarily the VCH applianceVM, but may also include port groups and other objects.
```
Comment (cgtexmex) [inline doc/design/vic-machine.md]: drop the "a" between against and vSphere...

File: usage.md [hunk 20]
```diff
@@ -1,6 +1,6 @@
 ## Installing Virtual Integrated Containers
 
-The intent is that vSphere Integrated Containers (VIC) should not _require_ an installation step - deploying a [Virtual Container Host](doc/arch/vic-container-abstraction.md#virtual-container-host) (VHC) directly without any prior steps should always be possible. At the current time this is the only approach available.
+The intent is that vSphere Integrated Containers (VIC) should not _require_ an installation step - deploying a [Virtual Container Host](doc/design/arch/vic-container-abstraction.md#virtual-container-host) (VHC) directly without any prior steps should always be possible. At the current time this is the only approach available.
```
Comment (cgtexmex) [inline usage.md]: s/VHC/VCH/

File: doc/design/components.md [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+* requires opening a port on the ESX firewall
+
+The alternative we're looking at is vSocket (uses PIO based VMCI communication), however that it Host<->VM only so we need a mechanism to relay that communication to the VCH. Initially it's expected that the Host->VCH communication still be a TCP connection for a staged delivery approach, with the longer term being an agent<->agent relay between the two hosts.
+
+[Issues relating to vsocket relay agent](https://github.com/vmware/vic/labels/component%2Fvsocket-relay-agent)
+
+
+### NSX Authenticating Agent
+
+As with [access to vSphere](#vmomi-authenticating-agent), NSX management interaction requires credentials and therefore credential management. The authenticating agent is indended to move the credential management problem out of the VCH and into the vSphere infrastructure. Authorizing the VCH to perform specific operations by virtue of being a specific VM (and potentially cryptographically verified), rather than generating and embedding credentials into the guest, means:
```
Comment (mlh78750) [inline doc/design/components.md]: s/idended/intended/

File: doc/design/configuration.md [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+
+Configures VCH on initial deploy:
+* _vSphere target and credentials_ (hidden)
+* networks (external, client, management, bridge, mapped vsphere networks)
+* vsphere compute path (e.g. resource pool)
+* image store URI
+* container store URI
+* permissible URIs for volumes
+* white/black list of registries
+* _etc_
```
Comment (mlh78750) [inline doc/design/configuration.md]: Is this also where you expect mapping of externally scoped networks will happen?

**Comment by mlh78750** [comment]
This is really well done. I put a few nit comments and a couple questions as comments in the diff. Address as you feel is appropriate and LGTM.

File: doc/design/components.md [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+
+Some of the major technical assumptions in the Docker codebase do not hold true for VIC (as of decision nexus date):
+
+1. container filesystems exist on the same host the daemon is running on
+2. namespaces can be created independently of containers (e.g. libnetwork sets up interfaces prior to graphdriver.Create)
+3. operations on namespaces are _relatively_ cheap in terms of performance
+4. container lifecycles are tied to that of the docker daemon
+5. no orthogonal second management plane
+
+As Docker continues with the decomposition of the monolithic docker daemon into components, our ability to consume those components directly increases, and there are very good reasons to do so - not least is the fact it reduces the develpment and maintenance burden on the VIC team. In line with the stated approach, as docker components become available that can be consumed, we will.
```
Comment (jzt) [inline doc/design/components.md]: s/develpment/development

File: doc/design/vic-machine.md (hunk 17 above)
Comment (hickeng) [inline doc/design/vic-machine.md]: This was actually intentional - this doc is one of the ways I'm communicating with Emma and I wanted a less definite phrase that "The current approach".
I've altered it to "Recommended initial approach" ...

File: doc/design/configuration.md (hunk 22 above)
Comment (hickeng) [inline doc/design/configuration.md]: yes - I view it as a core part of the VCH configuration. We definitely need to be able to update those mappings - preferably without needing to restart the appliance, and I expect reconfiguration to be done via vic-machine as well.

**Pull request merged**